components/cu_transform/src/velocity_cache.rs

```diff
@@ -0,0 +1,160 @@
+use crate::velocity::VelocityTransform;
+use cu29::clock::CuTime;
+use dashmap::DashMap;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+
+/// The cache entry for a velocity transform query
+#[derive(Clone)]
+pub(crate) struct VelocityTransformCacheEntry<T: Copy + Debug + 'static> {
+    /// The cached velocity transform result
+    pub(crate) velocity: VelocityTransform<T>,
+    /// The time for which this velocity was calculated
+    pub(crate) time: CuTime,
+    /// When this cache entry was last accessed
+    pub(crate) last_access: Instant,
+    /// Path hash used to calculate this velocity
+    pub(crate) path_hash: u64,
+}
+
+/// A cache for velocity transforms to avoid recalculating frequently accessed paths
+pub(crate) struct VelocityTransformCache<T: Copy + Debug + 'static> {
+    /// Map from (source, target) frames to cached velocity transforms
+    entries: DashMap<(String, String), VelocityTransformCacheEntry<T>>,
+    /// Maximum size of the cache
+    max_size: usize,
+    /// Maximum age of cache entries before invalidation
+    max_age: Duration,
+    /// Last time the cache was cleaned up (as nanoseconds since epoch)
+    last_cleanup: AtomicU64,
+    /// Cleanup interval - only clean every N seconds
+    cleanup_interval: Duration,
+}
+
+/// Get current time as nanoseconds since some epoch
+/// This is just used for comparing intervals, not for absolute time
+fn now_timestamp_nanos() -> u64 {
+    let now = Instant::now();
+    // We can't get nanoseconds from Instant directly, but we can use this trick
+    // to get a stable value that increases monotonically
+    let duration = now.elapsed();
+    (duration.as_secs() * 1_000_000_000 + duration.subsec_nanos() as u64) % u64::MAX
+}
+
+impl<T: Copy + Debug + 'static> VelocityTransformCache<T> {
+    pub(crate) fn new(max_size: usize, max_age: Duration) -> Self {
+        Self {
+            entries: DashMap::with_capacity(max_size),
+            max_size,
+            max_age,
+            last_cleanup: AtomicU64::new(now_timestamp_nanos()),
+            cleanup_interval: Duration::from_secs(5), // Clean every 5 seconds
+        }
+    }
+
+    /// Get a cached velocity transform if it exists and is still valid
+    pub(crate) fn get(
+        &self,
+        from: &str,
+        to: &str,
+        time: CuTime,
+        path_hash: u64,
+    ) -> Option<VelocityTransform<T>> {
+        let key = (from.to_string(), to.to_string());
+
+        if let Some(mut entry) = self.entries.get_mut(&key) {
+            let now = Instant::now();
```

Comment:
This cannot happen because of determinism. It has to be based on the robot clock otherwise the replay will be totally random. (imagine someone might go back in time, restore the state of the task and play again)